Source/Plugins/Tilemaps/Core/Tilemaps/TilemapFloodFill.cs

```diff
@@ -0,0 +1,213 @@
+
+						// Adjust the fill area
+						fillAreaSize.X += Math.Max(fillAreaTopLeft.X - pos.X, 0);
+						fillAreaTopLeft.X = Math.Min(fillAreaTopLeft.X, pos.X);
+
+						// If something above the current scan line is free, handle it recursively
+						if (pos.Y != 0 && FillAreaIsCandidate(fillBuffer, tiles, new Point2(pos.X, pos.Y - 1), baseTile))
+						{
+							// Find the topleft-most tile to start with
+							Point2 targetPos = new Point2(pos.X, pos.Y - 1);
+							targetPos = FillAreaFindTopLeft(fillBuffer, tiles, targetPos, baseTile);
+							if (!InternalFillArea(fillBuffer, tiles, targetPos, baseTile, ref maxTileCount, ref fillAreaTopLeft, ref fillAreaSize))
+								return false;
+						}
+					}
+				}
+
+				// Fill the current row
+				for (; rowPos.X < tiles.Width && FillAreaIsCandidate(fillBuffer, tiles, rowPos, baseTile); rowLength++, rowPos.X++)
+				{
+					fillBuffer[rowPos.X, rowPos.Y] = true;
+				}
+				maxTileCount -= rowLength;
+				if (maxTileCount < 0) return false;
+
+				// Adjust the fill area
+				fillAreaSize.X = Math.Max(fillAreaSize.X, rowPos.X - fillAreaTopLeft.X);
+
+				// If the current row is shorter than the previous, see if there are
+				// disconnected pixels below the (filled) previous row left to handle
+				if (rowLength < lastRowLength)
+				{
+					for (int end = pos.X + lastRowLength; ++rowPos.X < end;)
+					{
+						// Recursively handle the disconnected below-bottom pixels of the last row
+						if (FillAreaIsCandidate(fillBuffer, tiles, rowPos, baseTile))
+						{
+							if (!InternalFillArea(fillBuffer, tiles, rowPos, baseTile, ref maxTileCount, ref fillAreaTopLeft, ref fillAreaSize))
+								return false;
+						}
+					}
+				}
+				// If the current row is longer than the previous, see if there are
+				// top pixels above this one that are disconnected from the last row
+				else if (rowLength > lastRowLength && pos.Y != 0)
+				{
+					for (int prevRowX = pos.X + lastRowLength; ++prevRowX < rowPos.X;)
+					{
+						// Recursively handle the disconnected pixels of the last row
+						if (FillAreaIsCandidate(fillBuffer, tiles, new Point2(prevRowX, pos.Y - 1), baseTile))
+						{
+							// Find the topleft-most tile to start with
+							Point2 targetPos = new Point2(prevRowX, pos.Y - 1);
+							targetPos = FillAreaFindTopLeft(fillBuffer, tiles, targetPos, baseTile);
+							if (!InternalFillArea(fillBuffer, tiles, targetPos, baseTile, ref maxTileCount, ref fillAreaTopLeft, ref fillAreaSize))
+								return false;
+						}
+					}
+				}
+
+				lastRowLength = rowLength;
+				pos.Y++;
+
+				// Adjust the fill area
+				fillAreaSize.Y = Math.Max(fillAreaSize.Y, pos.Y - fillAreaTopLeft.Y);
+			}
+			while (lastRowLength != 0 && pos.Y < tiles.Height);
+
+			return true;
+		}
+
+		private Point2 FillAreaFindTopLeft(Grid<bool> fillBuffer, Grid<Tile> tiles, Point2 pos, Tile baseTile)
+		{
+			// Find the topleft-most connected matching tile
+			while (true)
```

Comment:
given that we have a test for completion shouldn't we rework this to have the test be up here?